cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -107,53 +112,154 @@ void copy_if_else_kernel(  column_device_view const lhs,
+                                       FilterFn filter,
+                                       rmm::mr::device_memory_resource *mr,
+                                       cudaStream_t stream)
+   {                
+      auto validity_policy = rhs.has_nulls() ? experimental::mask_allocation_policy::ALWAYS : experimental::mask_allocation_policy::NEVER;
+      std::unique_ptr<column> out = experimental::allocate_like(rhs, rhs.size(), validity_policy, mr);
+
+      // if length is 0, return immediately.  make_pair_iterator() does not like nullptr for element input
+      if(rhs.size() == 0){
+         return out;
+      }
+
+      auto out_dv = mutable_column_device_view::create(*out);
+      size_type out_null_count;
 
-         out->set_null_count(lhs.size() - valid_count.value());
+      // data and validity iterators for scalar
+      auto lhs_iter = thrust::make_constant_iterator(static_cast<cudf::experimental::scalar_type_t<Element>const*>(&lhs)->value());
+      auto lhs_valids = thrust::make_constant_iterator(1);
+      
+      if(validity_policy == experimental::mask_allocation_policy::ALWAYS){
+         launch(  thrust::make_zip_iterator(thrust::make_tuple(lhs_iter, lhs_valids)),
+                  make_pair_iterator<true, Element>((Element*)rhs.begin<Element>(), rhs.null_mask()),
+                  filter, *out_dv, out_null_count, mr, stream);
       } else {
-         // call the kernel
-         copy_if_else_kernel<block_size, Element, FilterFn, false><<<grid.num_blocks, block_size, 0, stream>>>(
-            *lhs_dv, *rhs_dv, filter, *out_dv, nullptr);
+         launch(  thrust::make_zip_iterator(thrust::make_tuple(lhs_iter, lhs_valids)),
+                  make_pair_iterator<false, Element>((Element*)rhs.begin<Element>(), static_cast<bitmask_type*>(nullptr)),
+                  filter, *out_dv, out_null_count, mr, stream);
       }
 
+      out->set_null_count(out_null_count);
+
+      return out;
+   }
+      
+   // scalar/scalar
+   std::unique_ptr<column> operator()( scalar const &lhs,
+                                       scalar const &rhs,
+                                       size_type out_size,
+                                       FilterFn filter,
+                                       rmm::mr::device_memory_resource *mr,
+                                       cudaStream_t stream)
+   {
+      std::unique_ptr<column> out = make_fixed_width_column(lhs.type(), out_size, UNALLOCATED, stream, mr);
+      auto out_dv = mutable_column_device_view::create(*out);
+      size_type out_null_count_unused;
+
+      auto all_valid = thrust::make_constant_iterator(1);
+
+      auto lhs_iter = thrust::make_constant_iterator(static_cast<cudf::experimental::scalar_type_t<Element>const*>(&lhs)->value());
+      auto rhs_iter = thrust::make_constant_iterator(static_cast<cudf::experimental::scalar_type_t<Element>const*>(&rhs)->value());
+
+      launch(  thrust::make_zip_iterator(thrust::make_tuple(lhs_iter, all_valid)),
```

Comment:
This zip iterator should be replaced with a `make_pair_iterator` that works on a scalar.